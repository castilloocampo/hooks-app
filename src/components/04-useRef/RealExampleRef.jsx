import { useState } from 'react';
import MultipleCustomHooks from '../03-examples/MultipleCustomHooks';


const RealExampleRef = () => {

  const [show, setShow] = useState(false);

  return (
    <div>
      <h1>RealExampleRef</h1>
      <hr />

      { show && <MultipleCustomHooks />}

      <button
        className='mt-5 inline-block align-middle text-center select-none border font-normal whitespace-no-wrap rounded py-1 px-3 leading-normal no-underline
        bg-blue-600 text-white active:bg-blue-900
        '
        onClick={()=>setShow(!show)} >
        Toggle Component
      </button>
    </div>
  );
}

export default RealExampleRef;
