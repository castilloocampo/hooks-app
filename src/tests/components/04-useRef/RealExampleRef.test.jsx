import RealExampleRef from '../../../components/04-useRef/RealExampleRef';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

describe('RealExampleRef', () => {
  test('should show correctly', () => {
    const { container } = render(<RealExampleRef />);
    expect(container).toMatchSnapshot();
  });

  test('should show <MultipleCustomHooks>', async () => {
    render(<RealExampleRef />);
    //Se carga por primera vez el componente
    const buttonMultiple = screen.getByTestId('button-multiplecustomhook');
    //Se hace cambio en propiedad
    fireEvent.click(buttonMultiple);
    //Se muestra componente por cambio de propiedad
    expect(screen.queryByText(/BreakingBad/)).toHaveTextContent("BreakingBad Quotes");

    let load = null, quote = null, buttonNext = null;
    //Se muestra elemento loading antes de mostrar el quote
    await waitFor(() => {
      load = screen.queryByTestId('loading');
      expect(load).toHaveTextContent(/Loading/);
    });
    //Se muestra el quote
    await waitFor(() => {
      load = screen.queryByTestId('loading');
      expect(load).toBe(null);
    });
    quote = screen.queryByTestId('blockquote').textContent
    buttonNext = screen.queryByTestId('button-next-quote');
      
    //Se Solicita mostrar nuevo elemento
    fireEvent.click(buttonNext);

    //Se muestra elemento loading antes de mostrar el nuevo quote
    await waitFor(() => {
      load = screen.queryByTestId('loading');
      expect(load).toHaveTextContent(/Loading/);
    });

    //Se muestra el nuevo quote
    await waitFor(() => {
      load = screen.queryByTestId('loading');
      expect(load).toBe(null);
      expect(screen.queryByTestId('blockquote').textContent).not.toBe(quote);
    });

    //Se hace cambio en propiedad
    fireEvent.click(buttonMultiple);

    //Se oculta componente por cambio de propiedad
    expect(screen.queryByText(/BreakingBad/)).toBe(null);
  });
})
