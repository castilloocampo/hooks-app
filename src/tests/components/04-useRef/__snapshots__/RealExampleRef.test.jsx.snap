// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RealExampleRef should show correctly 1`] = `
<div>
  <div>
    <h1>
      RealExampleRef
    </h1>
    <hr />
    <button
      class="mt-5 inline-block align-middle text-center select-none border font-normal whitespace-no-wrap rounded py-1 px-3 leading-normal no-underline bg-blue-600 text-white active:bg-blue-900 "
      data-testid="button-multiplecustomhook"
    >
      Toggle Component
    </button>
  </div>
</div>
`;
